tls/s2n_handshake_io.c

```diff
@@ -306,24 +306,40 @@ int s2n_conn_set_handshake_type(struct s2n_connection *conn)
 
 static int s2n_conn_update_handshake_hashes(struct s2n_connection *conn, struct s2n_blob *data)
 {
-    /* The handshake MD5 hash state will fail the s2n_hash_is_available() check
-     * since MD5 is not permitted in FIPS mode. This check will not be used as
-     * the handshake MD5 hash state is specifically used by the TLS 1.0 and TLS 1.1
-     * PRF, which is required to comply with the TLS 1.0 and 1.1 RFCs and is approved
-     * as per NIST Special Publication 800-52 Revision 1.
-     */
-    GUARD(s2n_hash_update(&conn->handshake.md5, data->data, data->size));
+    if (s2n_handshake_is_hash_required(&conn->handshake, S2N_HASH_MD5)) {
+        /* The handshake MD5 hash state will fail the s2n_hash_is_available() check
+         * since MD5 is not permitted in FIPS mode. This check will not be used as
+         * the handshake MD5 hash state is specifically used by the TLS 1.0 and TLS 1.1
+         * PRF, which is required to comply with the TLS 1.0 and 1.1 RFCs and is approved
+         * as per NIST Special Publication 800-52 Revision 1.
+         */
+        GUARD(s2n_hash_update(&conn->handshake.md5, data->data, data->size));
+    }
 
-    if (s2n_hash_is_available(S2N_HASH_MD5_SHA1)) {
-        /* The MD5_SHA1 hash cannot be initialized when FIPS mode is set. */
+    if (s2n_handshake_is_hash_required(&conn->handshake, S2N_HASH_SHA1)) {
+        GUARD(s2n_hash_update(&conn->handshake.sha1, data->data, data->size));
+    }
+
+    if (s2n_handshake_is_hash_required(&conn->handshake, S2N_HASH_MD5) &&
+        s2n_handshake_is_hash_required(&conn->handshake, S2N_HASH_SHA1)) {
         GUARD(s2n_hash_update(&conn->handshake.md5_sha1, data->data, data->size));
```

Comment:
I think this is going to break in FIPS mode. MD5 and SHA1 are required, but the `md5_sha1` hash state update will fail since it is not initialized. If you add the pre-existing `s2n_hash_is_available(S2N_HASH_MD5_SHA1)` check to this condition, it should work.

Reply:
updated